[pt][static_runtime] Memory model

hlu1 opened the pull request:
Summary:
The idea of the memory model is quite similar to that of BlackBoxPredictor, however, it's more complicated in pt due to 1) tensor views that share storage with storage refcount bumps but with different TensorImpls, 2) tensors sharing the same TensorImpl and the same storage, but with no refcount bump of the StorageImpl, 3) data types such as TensorList and Tuples that have Tensors in them, 4) need to support non-out/out variant mix while we move the aten ops to out variants.

As a result, I have to make the following adjustments:
1) remove tensors in output Tuples from internal blob list;
2) for memory allocation/deallocation, get candidate Tensors from the outputs of ops with out variant, extract StorageImpls from the Tensors, dedup, and remove output tensor StorageImpls, and get the final list of blobs for memory planning;
3) during the clean_up_memory pass, clean up memory held by the StorageImpls as well as Tensors/Lists/Tuples in IValues that don't participate in memory planning to reduce overall memory usage

Risk:
PyTorch team is planning to deprecate the current resize_outout api, which we do rely on. This is a pretty big risk.

https://www.internalfb.com/intern/diffusion/FBS/browsefile/master/fbcode/caffe2/aten/src/ATen/native/Resize.cpp?commit=6457b329847607553d34e788a3a7092f41f38895&lines=9-23

Test Plan:
```
buck test //caffe2/test:static_runtime
buck test //caffe2/benchmarks/static_runtime:static_runtime_cpptest
buck test //caffe2/caffe2/fb/predictor:pytorch_predictor_test
```
Benchmarks:
```
MKL_NUM_THREADS=1 OMP_NUM_THREADS=1 numactl -m 0 -C 13 \
buck-out/opt/gen/caffe2/caffe2/fb/predictor/ptvsc2_predictor_bench \
--scripted_model=/home/hlu/ads/adindexer/adindexer_ctr_mobilefeed/pt/merge/traced_precomputation.pt \
--pt_inputs=/home/hlu/ads/adindexer/adindexer_ctr_mobilefeed/pt/merge/container_precomputation_bs1.pt \
--iters=1000 --warmup_iters=10000 --num_threads=1 --pt_enable_static_runtime=true \
--pt_cleanup_activations=true --pt_enable_out_variant=false
```

|pt_cleanup_activations	|pt_enable_out_variant	|old ms/iter	|new ms/iter	|
|---	|---	|---	|---	|
|0	|0	|0.31873	|0.30228	|
|0	|1	|0.30018	|0.29184	|
|1	|0	|0.35246	|0.31895	|
|1	|1	|0.35742	|0.30417	|

Differential Revision: D24471854

bwasti commented on torch/csrc/jit/runtime/static/impl.h:
```diff
@@ -151,6 +163,25 @@ class TORCH_API StaticRuntime {
+
+ private:
+  std::unordered_set<size_t> reg_out_variant_;
+  std::vector<c10::StorageImpl*> internal_blob_storages_;
```
nit: `internal_storages_`? just since `blob` is an ever-so-slightly different concept and might be confusing

bwasti commented on torch/csrc/jit/runtime/static/impl.h:
```diff
@@ -151,6 +163,25 @@ class TORCH_API StaticRuntime {
     DCHECK(i < module_->output_regs.size());
     return reg_[module_->output_regs[i]];
   }
+  friend class MemoryPlanner;
```
let's avoid this by providing `nodes() const` and `inference_module() const` functions

bwasti commented on torch/csrc/jit/runtime/static/impl.h:
```diff
@@ -151,6 +163,25 @@ class TORCH_API StaticRuntime {
+  friend class MemoryPlanner;
+};
+
+class MemoryPlanner {
```
can you comment a quick overview about what this does (handles memory allocation at the graph level and provides two methods that are roughly `allocate` and `deallocate`)

hlu1 replied:
Will do

bwasti commented on torch/csrc/jit/runtime/static/impl.h:
```diff
@@ -151,6 +163,25 @@ class TORCH_API StaticRuntime {
+  MemoryPlanner(StaticRuntime* runtime);
+
+  void allocate_memory();
+  void clean_up_memory(std::vector<IValue>& reg);
```
we can cache a reference to `std::vector<IValue>& reg` in the constructor of MemoryPlanner such that we can make the signature `clean_up_memory()`.

It might also be cleaner to just name these `allocate()` and `deallocate()`

bwasti commented on torch/csrc/jit/runtime/static/impl.cpp:
```diff
@@ -309,12 +339,19 @@ StaticRuntime::IndividualMetrics StaticRuntime::benchmark_individual_ops(
 
   // main runs
   for (int i = 0; i < main_runs; i++) {
+    if (planner_) {
```
what about a default `TrivialMemoryPlanner` that basically no-ops the `allocate_memory` call?  it would require a bit of a refactor for the `deallocate_registers` so let's not do it in this PR, but I'm curious your opinion on the suggestion

hlu1 replied:
Why do you need to do that?

bwasti reviewed (approved):
this looks really good.  I think there are clear points for further extensions that can happen in a later diff (e.g. custom MemoryPlanners etc)

One thing I'd like to emphasize as a general comment is that as we add classes and attributes to those classes we'll have to start being very careful with naming.  We already have a slew of Caffe2 and PyTorch concepts floating around that might get a bit confusing in the future.

dzhulgakov commented on torch/csrc/jit/runtime/static/impl.cpp:
```diff
@@ -27,6 +27,25 @@ void CheckGraphEligibility(const std::shared_ptr<torch::jit::Graph>& graph) {
       throw std::runtime_error("Cannot accelerate unfrozen graphs");
     }
   }
+  // We support output types include None, Tensor and List or Tuple of Tensor
```
you probably want to give a bit friendlier error message, at least print the output type and say what it's allowed to be

dzhulgakov commented on torch/csrc/jit/runtime/static/impl.cpp:
```diff
@@ -333,6 +370,127 @@ StaticRuntime::IndividualMetrics StaticRuntime::benchmark_individual_ops(
+
+    internal_blob_max_sizes_.resize(internal_blob_storages_.size());
+    reg_out_variant_.clear(); // no longer needed
+  }
```
you might want to add some sanity check here that Storages don't change between iterations. General check is probably expensive, but you can run it in debug mode only or something (or on second iteration only). Basically if there's a bug somewhere in how we handle output argument, then reg_out_variant's element would get a new tensor storage assigned each time which won't be present in internal_blob_storages_.

hlu1 replied:
Will do

dzhulgakov commented on torch/csrc/jit/runtime/static/ops.cpp:
```diff
@@ -98,7 +104,14 @@ getOutOfPlaceOperation(Node* n) {
+}
+
+std::function<void(const ProcessedNode*, std::vector<IValue>&)>
+getNativeOperation(Node* n) {
```
are these all "view producing" operators? add a comment/todo to extend them in the future (I think there are about 10 of those if I recall correctly: e.g. view, narrow, etc - we must have a list somewhere in jit already)

hlu1 replied:
Yes, in most of the cases they return a Tensor view, though it can return a copy of `self` if they're the same:

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorShape.cpp#L1617-L1619

Another corner case is when the input tensor is mkldnn. Static Runtime does not support this case yet.

dzhulgakov commented on torch/csrc/jit/runtime/static/ops.cpp:
```diff
@@ -4,26 +4,32 @@
+}
+
+bool canRunNatively(Node* n) {
+  static std::unordered_set<std::string> out_of_place_nodes{"aten::transpose",
```
nit: rename the constant?

hlu1 replied:
Nice catch

hlu1 commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -504,6 +504,9 @@ std::ostream& IValue::repr(
       return out << enum_holder->qualifiedClassName() << "." <<
           enum_holder->name();
     }
+    case IValue::Tag::Object:
```
My diff was stacked on top of #45556 and shipit merged these two during diff exporting.

hlu1 committed:
[pt][static_runtime] Memory model (#46896)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/46896

The idea of the memory model is quite similar to that of BlackBoxPredictor, however, it's more complicated in pt due to 1) tensor views that share storage with storage refcount bumps but with different TensorImpls, 2) tensors sharing the same TensorImpl and the same storage, but with no refcount bump of the StorageImpl, 3) data types such as TensorList and Tuples that have Tensors in them, 4) need to support non-out/out variant mix while we move the aten ops to out variants.

As a result, I have to make the following adjustments:
1) remove tensors in output Tuples from internal blob list;
2) for memory allocation/deallocation, get candidate Tensors from the outputs of ops with out variant, extract StorageImpls from the Tensors, dedup, and remove output tensor StorageImpls, and get the final list of blobs for memory planning;
3) during the clean_up_memory pass, clean up memory held by the StorageImpls as well as Tensors/Lists/Tuples in IValues that don't participate in memory planning to reduce overall memory usage

Risk:
PyTorch team is planning to deprecate the current resize_outout api, which we do rely on. This is a pretty big risk.

https://www.internalfb.com/intern/diffusion/FBS/browsefile/master/fbcode/caffe2/aten/src/ATen/native/Resize.cpp?commit=6457b329847607553d34e788a3a7092f41f38895&lines=9-23

Test Plan:
```
buck test //caffe2/test:static_runtime
buck test //caffe2/benchmarks/static_runtime:static_runtime_cpptest
buck test //caffe2/caffe2/fb/predictor:pytorch_predictor_test
```
Benchmarks:
```
MKL_NUM_THREADS=1 OMP_NUM_THREADS=1 numactl -m 0 -C 13 \
buck-out/opt/gen/caffe2/caffe2/fb/predictor/ptvsc2_predictor_bench \
--scripted_model=/home/hlu/ads/adindexer/adindexer_ctr_mobilefeed/pt/merge/traced_precomputation.pt \
--pt_inputs=/home/hlu/ads/adindexer/adindexer_ctr_mobilefeed/pt/merge/container_precomputation_bs1.pt \
--iters=1000 --warmup_iters=10000 --num_threads=1 --pt_enable_static_runtime=true \
--pt_cleanup_activations=true --pt_enable_out_variant=false
```

|pt_cleanup_activations	|pt_enable_out_variant	|old ms/iter	|new ms/iter	|
|---	|---	|---	|---	|
|0	|0	|0.31873	|0.30228	|
|0	|1	|0.30018	|0.29184	|
|1	|0	|0.35246	|0.31895	|
|1	|1	|0.35742	|0.30417	|

Differential Revision: D24471854

fbshipit-source-id: 5991ae61e41e5c4e663dadff0995dd1c53cd7ee6
